tools/k8s_versions/update_k8s_versions.go

```diff
@@ -0,0 +1,238 @@
+func updateK8sVersion(filePath string, kubernetesVersions []KubernetesVersion) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+	defer file.Close()
+
+	var output []string
+	var k8sVersionBlock []string
+	var inK8sVersionBlock bool
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		trimmed := strings.TrimSpace(line)
+
+		// Locate the k8s-version block needed to update
+		if strings.HasPrefix(trimmed, "k8s-version:") || strings.HasPrefix(trimmed, "kubernetes_version:") {
+			inK8sVersionBlock = true
+			output = append(output, line)
+			continue
+		}
+		// if we found the k8s version array, extract all versions and pass them to updateVersionsBlock
+		if inK8sVersionBlock {
+			if strings.HasPrefix(trimmed, "-") || strings.HasPrefix(trimmed, "#") {
+				k8sVersionBlock = append(k8sVersionBlock, line)
+				continue
+			} else if trimmed == "" || !strings.HasPrefix(trimmed, "-") {
+				updatedVerBlock, err := updateVersionsBlock(kubernetesVersions, k8sVersionBlock)
+				if err != nil {
+					return err
+				}
+				output = append(output, append(updatedVerBlock, line)...)
+				inK8sVersionBlock = false
+				continue
+			}
+		}
+		output = append(output, line)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("failed to read file: %w", err)
+	}
+
+	if err := os.WriteFile(filePath, []byte(strings.Join(output, "\n")+"\n"), 0644); err != nil {
+		return fmt.Errorf("failed to write updated file: %w", err)
+	}
+	return nil
+}
+
+func updateVersionsBlock(versions []KubernetesVersion, versionBlock []string) ([]string, error) {
+	var output []string
+	now := time.Now()
+	const layout = "2006-01-02"
+	versionRE := regexp.MustCompile("\\d+.\\d+.\\d+")
+
+	for _, version := range versions {
+		eolDate, err := time.Parse(layout, version.EOLDate)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing date: %w", err)
+		}
+
+		for i, line := range versionBlock {
+			trimmed := strings.TrimSpace(line)
+			tag := ""
+			if strings.HasPrefix(trimmed, fmt.Sprintf("- v%s", version.Cycle)) {
+				if eolDate.After(now) {
+					tag, err = getAvailableKindImage(version.Latest)
+					if err != nil {
+						return nil, fmt.Errorf("failed to get available kind image: %w", err)
+					}
+					line = versionRE.ReplaceAllString(line, tag)
+					output = append(output, line)
+				}
+				break
+			} else if i == len(versionBlock)-1 && eolDate.After(now) {
+				// Add the new version at the end of the list
+				tag, err = getAvailableKindImage(version.Latest)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get available kind image: %w", err)
+				}
+				output = append(output, fmt.Sprintf("%s- v%s # EOL %s", strings.Repeat(" ", len(line)-len(trimmed)), tag, version.EOLDate))
+			}
+		}
+	}
+	return output, nil
+}
+
+func getAvailableKindImage(tag string) (string, error) {
```

Comment:
This logic is pretty hard for me to follow. It looks like for every k8s version we're getting an available kind image. Is there some way to iterate over the kind images supported, rather than manually doing string operations to deprecate versions and check if they're valid? If we [ordered results simply by name](https://hub.docker.com/v2/repositories/kindest/node/tags?=page=1&ordering=name), we'd be able to just iterate directly checking if the k8s version exists as a kind version, and if not, keep iterating. We could also use [go-version](https://pkg.go.dev/github.com/hashicorp/go-version) to handle some of the versioning string/regex functionality comparing values if it's useful.

Reply:
This was also drastically simplified in the next iteration.
But the idea of getting the kind images first is not ideal, as you will also have to do with paging and large payloads as there are many images/tags for different platform etc...
The idea here and still in the next iteration, that we have a list of "currently" supported k8s versions, with latest and cycle (those are normally 3 or 4 max versions), we iterate through the latest and decrement it until we find an existing image on docker hub and use that tag.